app/controllers/api_controller/parameters.rb

```diff
@@ -22,44 +22,51 @@ def hash_fetch(hash, element, default = {})
     end
 
     #
-    # Returns an Arel definition for ActiveRecord's option for where()
-    # based on the filter attributes specified.
+    # Returns an MiqExpression based on the filter attributes specified.
     #
     def filter_param(klass)
       return nil if params['filter'].blank?
 
-      operators = {"!=" => {:default => :not_eq, :str => :does_not_match, :ruby => :!=},
-                   "<=" => {:default => :lteq},
-                   ">=" => {:default => :gteq},
-                   "<"  => {:default => :lt},
-                   ">"  => {:default => :gt},
-                   "="  => {:default => :eq, :str => :matches, :ruby => :==}
-                  }
+      operators = {
+        "!=" => {:default => "!=", :regex => "REGULAR EXPRESSION DOES NOT MATCH", :null => "IS NOT NULL"},
+        "<=" => {:default => "<="},
+        ">=" => {:default => ">="},
+        "<"  => {:default => "<"},
+        ">"  => {:default => ">"},
+        "="  => {:default => "=", :regex => "REGULAR EXPRESSION MATCHES", :null => "IS NULL"}
+      }
 
-      res_filter = nil
-      ruby_filters = []
+      and_expressions = []
+      or_expressions = []
 
       params['filter'].select(&:present?).each do |filter|
         parsed_filter = parse_filter(filter, operators)
-        if klass.column_names.include?(parsed_filter[:attr])
-          arel = klass.arel_table[parsed_filter[:attr]].send(parsed_filter[:method], parsed_filter[:value])
-          res_filter = if res_filter.nil?
-                         arel
-                       else
-                         parsed_filter[:logical_or] ? res_filter.or(arel) : res_filter.and(arel)
-                       end
-        elsif parsed_filter[:ruby_operator].present? && klass.virtual_attribute?(parsed_filter[:attr])
-          ruby_filters << lambda do |result_set|
-            result_set.select do |resource|
-              attr = resource.public_send("#{parsed_filter[:attr]}")
-              attr.send("#{parsed_filter[:ruby_operator]}", parsed_filter[:value])
-            end
-          end
-        else
-          raise BadRequestError, "attribute #{parsed_filter[:attr]} does not exist"
+        parts = parsed_filter[:attr].split(".")
+        attr = parts.pop
+        unless virtual_or_physical_attribute?(target_class(klass, parts), attr)
```

Comment:
Found a way to generalize this for all supported cases, though wonder if this needs some special error handling, @abellotti ?